File: common/src/main/java/io/xol/chunkstories/world/region/RegionImplementation.java
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package io.xol.chunkstories.world.region;

import java.io.File;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;
import java.util.Random;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicLong;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

import io.xol.chunkstories.api.entity.Entity;
import io.xol.chunkstories.api.util.CompoundIterator;
import io.xol.chunkstories.api.util.IterableIterator;
import io.xol.chunkstories.api.world.WorldMaster;
import io.xol.chunkstories.api.world.WorldUser;
import io.xol.chunkstories.api.world.chunk.Chunk;
import io.xol.chunkstories.api.world.chunk.ChunksIterator;
import io.xol.chunkstories.api.world.region.Region;
import io.xol.chunkstories.util.concurrency.SafeWriteLock;
import io.xol.chunkstories.world.WorldImplementation;
import io.xol.chunkstories.world.chunk.ChunkHolderImplementation;
import io.xol.chunkstories.world.chunk.CompressedData;
import io.xol.chunkstories.world.chunk.CubicChunk;
import io.xol.chunkstories.world.io.IOTask;
import io.xol.chunkstories.world.region.format.CSFRegionFile;

public class RegionImplementation implements Region {
	public final WorldImplementation world;
	public final int regionX, regionY, regionZ;
	public final long uuid;

	protected Collection<CubicChunk> loadedChunks = ConcurrentHashMap.newKeySet();

	/**
	 * Contains the users that registered to this region using world.aquireRegion or
	 * region.registerUser
	 */
	public final Set<WorldUser> users = new HashSet<>();

	/**
	 * Keeps track of the number of users of this region. Warning: this includes
	 * users of not only the region itself, but ALSO the chunks that make up this
	 * region. An user registered in 3 chunks for instance, will be counted three
	 * times. When this counter reaches zero, the region is unloaded.
	 */
	public int usersCount = 0;

	/**
	 * Lock shared with the chunk holders to ensure consistency for the above fields
	 */
	public final Lock usersLock = new ReentrantLock();

	// Only relevant on Master worlds
	// public final CSFRegionFile handler;
	public final File file;
	public CSFRegionFile handler;

	private AtomicLong unloadCooldown = new AtomicLong();
	private boolean unloadedFlag = false;

	// TODO find a clean way to let IOTaks fiddle with this
	public SafeWriteLock chunksArrayLock = new SafeWriteLock();

	// Holds 8x8x8 CubicChunks
	private ChunkHolderImplementation[] chunkHolders;
	private AtomicBoolean isDiskDataLoaded = new AtomicBoolean(false);

	private static Random random = new Random();

	public RegionImplementation(WorldImplementation world, int regionX, int regionY, int regionZ) {
		this.world = world;
		this.regionX = regionX;
		this.regionY = regionY;
		this.regionZ = regionZ;

		if (regionX < 0 || regionY < 0 || regionZ < 0)
			throw new RuntimeException("Regions aren't allowed negative coordinates.");

		// Initialize slots
		chunkHolders = new ChunkHolderImplementation[512];
		for (int i = 0; i < 8; i++)
			for (int j = 0; j < 8; j++)
				for (int k = 0; k < 8; k++)
					chunkHolders[i * 64 + j * 8 + k] = new ChunkHolderImplementation(this, loadedChunks,
							regionX * 8 + i, regionY * 8 + j, regionZ * 8 + k);

		// Unique UUID
		uuid = random.nextLong();

		// Set the initial cooldown delay
		unloadCooldown.set(System.currentTimeMillis());

		// Only the WorldMaster has a concept of files
		if (world instanceof WorldMaster) {
			file = new File(world.getFolderPath() + "/regions/" + regionX + "." + regionY + "." + regionZ + ".csf");
			// CSFRegionFile.determineVersionAndCreate(this);
			world.getIoHandler().requestRegionLoad(this);
		} else {
			file = null;
			isDiskDataLoaded.set(true);
		}
	}

	@Override
	public IterableIterator<WorldUser> getUsers() {
		return new IterableIterator<WorldUser>() {
			Iterator<WorldUser> i = users.iterator();

			@Override
			public boolean hasNext() {
				return i.hasNext();
			}

			@Override
			public WorldUser next() {
				return i.next();
			}

		};
	}

	@Override
	public boolean registerUser(WorldUser user) {
		try {
			usersLock.lock();
			boolean notRedundant = users.add(user);

			if (notRedundant)
				usersCount++;

			return notRedundant;
		} finally {
			usersLock.unlock();
		}
	}

	@Override
	/**
	 * Unregisters user and if there is no remaining user, unloads the chunk
	 */
	public boolean unregisterUser(WorldUser user) {
		try {
			usersLock.lock();
			if (users.remove(user))
				usersCount--;

			if (usersCount == 0) {
				internalUnload();
				return true;
			}

			return false;
		} finally {
			usersLock.unlock();
		}

	}

	public int countUsers() {
		return usersCount;
	}

	public CompressedData getCompressedData(int chunkX, int chunkY, int chunkZ) {
		return chunkHolders[(chunkX & 7) * 64 + (chunkY & 7) * 8 + (chunkZ & 7)].getCompressedData();
	}

	@Override
	public CubicChunk getChunk(int chunkX, int chunkY, int chunkZ) {
		return chunkHolders[(chunkX & 7) * 64 + (chunkY & 7) * 8 + (chunkZ & 7)].getChunk();
	}

	@Override
	public ChunkHolderImplementation getChunkHolder(int chunkX, int chunkY, int chunkZ) {
		return chunkHolders[(chunkX & 7) * 64 + (chunkY & 7) * 8 + (chunkZ & 7)];
	}

	@Override
	public boolean isChunkLoaded(int chunkX, int chunkY, int chunkZ) {
		return chunkHolders[(chunkX & 7) * 64 + (chunkY & 7) * 8 + (chunkZ & 7)].isChunkLoaded();
	}

	public ChunksIterator iterator() {
		return new ChunksIterator() {

			Iterator<CubicChunk> i = loadedChunks.iterator();

			@Override
			public boolean hasNext() {
				return i.hasNext();
			}

			@Override
			public Chunk next() {
				return i.next();
			}

		};
	}

	@Override
	public boolean isDiskDataLoaded() {
		return isDiskDataLoaded.get();
	}

	public boolean isUnloaded() {
		return unloadedFlag;
	}

	public void setDiskDataLoaded(boolean b) {
		isDiskDataLoaded.set(b);
	}

	public void internalUnload() {
		if (/* should save */ world instanceof WorldMaster) {
			this.save();
		}

		// Before unloading the holder we want to make sure we finish all saving
		// operations
		// if (handler != null)
		// handler.finishSavingOperations();

		// Set unloaded flag to true so we are not using again an unloaded holder
		unloadedFlag = true;

		// Remove the reference in the world to this
		this.getWorld().getRegionsHolder().removeRegion(this);
	}

	@Override
	public IOTask save() {
		return world.getIoHandler().requestRegionSave(this);
	}

	/*
	 * @Override public IOTask unloadAndSave() { unload(); return
	 * world.ioHandler.requestRegionSave(this); }
	 */

	@Override
	public String toString() {
		return "[Region rx:" + regionX + " ry:" + regionY + " rz:" + regionZ + " uuid: " + uuid + "loaded?:"
				+ isDiskDataLoaded.get() + " u:" + unloadedFlag + " chunks: " + "NULL" + " entities:" + "tbf" + "]";
	}

	public void compressAll() {
		for (int a = 0; a < 8; a++)
			for (int b = 0; b < 8; b++)
				for (int c = 0; c < 8; c++)
					chunkHolders[a * 64 + b * 8 + c].compressChunkData();
	}

	public void compressChangedChunks() {
		for (int a = 0; a < 8; a++)
			for (int b = 0; b < 8; b++)
				for (int c = 0; c < 8; c++) {
					if (chunkHolders[a * 64 + b * 8 + c].getChunk() != null) {
						CubicChunk chunk = chunkHolders[a * 64 + b * 8 + c].getChunk();

						if (chunk.compr_uncomittedBlockModifications.get() > 0)
							// if (chunk.lastModification.getVoxelComponent() > chunk.lastModificationSaved.getVoxelComponent())
							chunkHolders[a * 64 + b * 8 + c].compressChunkData();
					}
				}
	}

	@Override
	public int getNumberOfLoadedChunks() {
		int count = 0;

		for (int a = 0; a < 8; a++)
			for (int b = 0; b < 8; b++)
				for (int c = 0; c < 8; c++)
					if (chunkHolders[a * 64 + b * 8 + c].isChunkLoaded())
						count++;

		return count;
	}

	@Override
	public int getRegionX() {
		return regionX;
	}

	@Override
	public int getRegionY() {
		return regionY;
	}

	@Override
	public int getRegionZ() {
		return regionZ;
	}

	@Override
	public IterableIterator<Entity> getEntitiesWithinRegion() {
		List<Iterator<Entity>> listOfIterators = new ArrayList<Iterator<Entity>>();
		for (int a = 0; a < 8; a++)
			for (int b = 0; b < 8; b++)
				for (int c = 0; c < 8; c++) {
					Chunk chunk = this.getChunk(a, b, c);
					if (chunk != null)
						listOfIterators.add(chunk.getEntitiesWithinChunk());
				}

		return new CompoundIterator<Entity>(listOfIterators);
	}

	public void resetUnloadCooldown() {
		unloadCooldown.set(System.currentTimeMillis());
	}

	public boolean canBeUnloaded() {
		// Don't unload it until it has been loaded for 10s
		return this.isDiskDataLoaded() && (System.currentTimeMillis() - this.unloadCooldown.get() > 10 * 1000L);
	}

	public WorldImplementation getWorld() {
		return world;
	}

	/**
	 * Unloads unused chunks, returns true if all chunks were unloaded
	 */
	/*
	 * public boolean unloadsUnusedChunks() { int loadedChunks = 0;
	 * 
	 * for (int a = 0; a < 8; a++) for (int b = 0; b < 8; b++) for (int c = 0; c <
	 * 8; c++) { chunkHolders[a * 64 + b * 8 + c].unloadsIfUnused();
	 * if(chunkHolders[a * 64 + b * 8 + c].isChunkLoaded()) loadedChunks++; }
	 * 
	 * return loadedChunks == 0; }
	 */

	// @Override
	/**
	 * Returns true if no one uses the region or one of it's chunk holders
	 */
	/*
	 * public boolean isUnused() { int usedChunks = 0;
	 * 
	 * for (int a = 0; a < 8; a++) for (int b = 0; b < 8; b++) for (int c = 0; c <
	 * 8; c++) { chunkHolders[a * 64 + b * 8 + c].unloadsIfUnused();
	 * if(chunkHolders[a * 64 + b * 8 + c].isChunkLoaded() || chunkHolders[a * 64 +
	 * b * 8 + c].countUsers() > 0) usedChunks++; }
	 * 
	 * //if(this.regionY == 0) // System.out.println(usedChunks + " vs " +
	 * this.countUsers());
	 * 
	 * return usedChunks == 0 && this.countUsers() == 0; }
	 */
}

File: common/src/main/java/io/xol/chunkstories/world/storage/RegionImplementation.kt
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package io.xol.chunkstories.world.storage

import io.xol.chunkstories.api.entity.Entity
import io.xol.chunkstories.api.world.WorldMaster
import io.xol.chunkstories.api.world.WorldUser
import io.xol.chunkstories.api.world.heightmap.Heightmap
import io.xol.chunkstories.api.world.region.Region
import io.xol.chunkstories.world.WorldImplementation
import io.xol.chunkstories.world.chunk.CubicChunk
import io.xol.chunkstories.world.io.IOTaskLoadRegion
import io.xol.chunkstories.world.io.IOTaskSaveRegion
import io.xol.chunkstories.world.region.format.CSFRegionFile
import org.slf4j.LoggerFactory
import java.io.File
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.locks.ReentrantLock

class RegionImplementation(override val world: WorldImplementation, override val heightmap: Heightmap, override val regionX: Int, override val regionY: Int, override val regionZ: Int) : Region, WorldUser {
    val file: File?
    val handler: CSFRegionFile?

    val stateLock = ReentrantLock()
    override var state: Region.State = Region.State.Zombie
        set(value) {
            field = value
            //println("DEBUG: Region ${this@RegionImplementation} state set to $value")
        }

    /** Keeps track of the number of users of this region. Warning: this includes users of not only the region itself, but ALSO the chunks that make up this
    region. An user registered in 3 chunks for instance, will be counted three times. When this counter reaches zero, the region is unloaded.*/
    var usersCount = 0
    internal val usersSet: MutableSet<WorldUser> = HashSet()
    /** Externally exposed read-only copy of the users set */
    override var users: Set<WorldUser> = emptySet()
    //val usersLock: Lock = ReentrantLock()

    private val chunkHolders: Array<ChunkHolderImplementation>

    internal var loadedChunksSet = ConcurrentHashMap.newKeySet<CubicChunk>()
    override val loadedChunks: Sequence<CubicChunk>
        get() = loadedChunksSet.asSequence()

    override val entitiesWithinRegion: Sequence<Entity>
        get() = loadedChunks.flatMap { it.localEntities.asSequence() }

    init {
        if (regionX < 0 || regionY < 0 || regionZ < 0)
            throw RuntimeException("Regions aren't allowed negative coordinates.")

        heightmap.registerUser(this)

        // Initialize slots
        chunkHolders = Array(512) { i ->
            val cx = (i shr 6) and 0x7
            val cy = (i shr 3) and 0x7
            val cz = (i shr 0) and 0x7
            //println("cx $cx cy $cy cz $cz")
            ChunkHolderImplementation(this, (regionX shl 3) + cx, (regionY shl 3) + cy, (regionZ shl 3) + cz)
        }

        // Only the WorldMaster has a concept of files
        if (world is WorldMaster) {
            file = File(world.folderPath + "/regions/" + regionX + "." + regionY + "." + regionZ + ".csf")
            handler = CSFRegionFile.determineVersionAndCreate(this)

            if (file.exists()) {
                //assert(heightmap.state !is Heightmap.State.Generating) { "We should never have existing data if the heightmap is still generating!" }

                val task = IOTaskLoadRegion(this)
                state = Region.State.Loading(task)
                world.ioHandler.scheduleTask(task)
            } else {
                state = Region.State.Available()
                chunkHolders.forEach { it.whenRegionIsAvailable() }
            }

        } else {
            // Remote-world regions don't wait for anything to load in
            file = null
            handler = null
            state = Region.State.Available()
            chunkHolders.forEach { it.whenRegionIsAvailable() }
        }
    }

    fun whenDataLoadedCallback() {
        try {
            stateLock.lock()
            if (state !is Region.State.Loading) {
                logger.error("Illegal state transition: When accepting loaded data, region was in state $state")
                return
            }

            state = Region.State.Available()
            chunkHolders.forEach { it.whenRegionIsAvailable() }
        } finally {
            stateLock.unlock()
        }
    }

    override fun registerUser(user: WorldUser): Boolean {
        try {
            //usersLock.lock()
            stateLock.lock()

            if(state is Region.State.Zombie)
                throw Exception("Registering user in a zombie region !!!")

            if (usersSet.add(user)) {
                usersCount++
                return true
            }
            return false
        } finally {
            users = usersSet.toSet()
            //usersLock.unlock()
            stateLock.unlock()
        }
    }

    override fun unregisterUser(user: WorldUser): Boolean {
        try {
            //usersLock.lock()
            stateLock.lock()
            usersSet.remove(user)
            usersCount--

            if (usersCount == 0) {
                unload()
                return true
            }

            return false
        } finally {
            users = usersSet.toSet()
            stateLock.unlock()
            //usersLock.unlock()
        }

    }

    override fun getChunk(chunkX: Int, chunkY: Int, chunkZ: Int): CubicChunk? {
        return chunkHolders[(chunkX and 7) * 64 + (chunkY and 7) * 8 + (chunkZ and 7)].chunk
    }

    override fun getChunkHolder(chunkX: Int, chunkY: Int, chunkZ: Int): ChunkHolderImplementation {
        return chunkHolders[(chunkX and 7) * 64 + (chunkY and 7) * 8 + (chunkZ and 7)]
    }

    /** Called both when the last user logs off and when saving is done */
    internal fun unload() {
        fun remove() {
            // Remove the reference in the world to this
            world.regionsStorage.removeRegion(this)
            heightmap.unregisterUser(this)

            state = Region.State.Zombie
        }

        try {
            stateLock.lock()

            if (state is Region.State.Loading) {
                val task = (state as Region.State.Loading).fence as IOTaskLoadRegion

                // If we managed to kill the task before it executes or before it could callback whenDataLoadedCallback
                if(task.tryCancel() || state !is Region.State.Available) {
                    remove()
                    return
                } else {
                    println("failed to cancel task :(, resulting state is $state")
                }
            }

            if (state !is Region.State.Available && state !is Region.State.Saving)
                throw Exception("Illegal state transition: When unloading, region was in state $state")

            // We want to save, but only once.
            // After the first save we'll still be in Saving state and so we can transition to Zombie.
            if (world is WorldMaster && state is Region.State.Available && state !is Region.State.Saving) {
                val task = IOTaskSaveRegion(this)
                state = Region.State.Saving(task)
                world.ioHandler.scheduleTask(task)

                return // we'll actually unload later
            }

            remove()

        } finally {
            stateLock.unlock()
        }
    }

    fun whenSavingDone() {
        try {
            stateLock.lock()
            if (usersCount == 0)
                unload()
            else {
                try {
                    //stateLock.lock()
                    if (state is Region.State.Saving)
                        state = Region.State.Available()
                } finally {
                    //stateLock.unlock()
                }
            }

        } finally {
            stateLock.unlock()
        }
    }

    override fun toString(): String {
        return ("[Region rx:$regionX ry:$regionY rz:$regionZ state:${state.javaClass.simpleName} chunks:${loadedChunks.count()} entities:${entitiesWithinRegion.count()}]")
    }

    fun compressAll() {
        for (a in 0..7)
            for (b in 0..7)
                for (c in 0..7)
                    chunkHolders[a * 64 + b * 8 + c].compressChunkData()
    }

    fun compressChangedChunks() {
        for (a in 0..7)
            for (b in 0..7)
                for (c in 0..7) {
                    val chunk = chunkHolders[a * 64 + b * 8 + c].chunk
                    if (chunk != null) {

                        if (chunk.compressionUncommitedModifications.get() > 0)
                            chunkHolders[a * 64 + b * 8 + c].compressChunkData()
                    }
                }
    }

    companion object {
        internal var logger = LoggerFactory.getLogger("world.storage")
    }
}
